File: 1일1알고/구현/0202/문자열압축.py
# def solution(s):
#     res = len(s)
#     letter = ''
#
#     for i in range(1, len(s) // 2 + 1):
#         cnt = 1
#         answer = ''
#         flag = False
#         for j in range(0, len(s), i):
#
#             if s[j:j+i] == s[j+i:j+(2*i)]:
#                 flag=True
#                 letter = s[j:j+i]
#                 cnt += 1
#                 continue
#             else:
#                 if not flag:
#                     answer += s[j:j+i]
#
#             if cnt>1:
#
#                 answer += (str(cnt)+letter)
#                 letter = ''
#                 cnt = 1
#                 flag = False
#         if res > len(answer):
#             res = len(answer)
#     return res



def solution(s):
    answer = len(s)
    #1개 단위 step 부터 압축 단위를 늘려가며 확인
    for step in range(1, len(s)//2+1):
        compressed = ''
        prev = s[0:step] #앞에서부터 step만큼의 문자열 추출
        count = 1
        for j in range(step, len(s), step):
            #이전 상태와 동일하다면 압축 횟수 증가
            if prev == s[j:j+step]:
                count += 1
            else:
                # 다른 문자열이 나왔다면 더이상 압축 x
                #A if 조건 else B
                #조건값이 True(참)이면 A를 False(거짓)이면 B를 반환
                compressed += str(count) + prev if count >=2 else prev
                #다시 상태 초기화
                prev = s[j:j+step]
                count = 1
        #남아있는 문자열에 대한 처리
        compressed += str(count) + prev if count >=2 else prev
        answer = min(answer,len(compressed))
    return answer

print(solution('ababcccab'))

File: 1일1알고/구현/0202/문자열압축.js


function solution(s) {
    console.log(s.length)
    console.log(s.length/2,'문자열최대길이')
    let mmin = Infinity
    let answer
    for (let i=1; i <= parseInt(s.length/2); i++) {
        console.log(i,'i')
        let cnt = 1
        answer = 0
        let flag
        let letter = ''
        for (let j=0; j < s.length-i; j+=i) {
            console.log(j,'j')
            console.log(s.substring(j,j+i),'좌')
            console.log(s.substring(j+i,j+2*i),'우')
            
            if (s.substring(j,j+i) === s.substring(j+i,j+2*i)) {
                flag = true
                letter = s.substring[j,j+i]
                cnt += 1
            } else {
                // flag = false
            }
            
            if (cnt>1) {
                answer += cnt.toString()+letter
                cnt = 1
                flag = false
            }
        }
        console.log(answer,'비교비교')
        if (answer<mmin) {
            mmin = answer
        }
    }

    return answer;
}